### PR TITLE
[NEW] Content-Security-Policy for inline scripts

### LINKED ISSUE #19815: Not possible to set an effective Content-Security-Policy (CSP) because of in-line content
### Description:

Background information needed to understand this bug-report can be found on mozilla blog: https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP (Content-Security-Policy = CSP)
[...] A CSP compatible browser will then only execute scripts loaded in source files received from those allowlisted domains, ignoring all other script (including inline scripts and event-handling HTML attributes).
CSP is a method to protect against XSS attacks. Rocket Chat is a chat client and are likely to be targeted with XSS attacks in the future. Microsoft Teams just had an incident with this which resulted in a non click remote code execution.

When visiting a rocket chat server, there's several "in-line" scripts and style content being served to the client. This is bad practice and kills the opportunity to set a CSP.

All of the in-line content should be moved to files, such as .js, .css or other equivalent files.

### Steps to reproduce:

1. Set HTTP header Content-Security-Policy "default-src 'self';"
2. Verify that header are set with https://securityheaders.com/
3. Open rocket chat URL using a CSP aware browser, such as Firefox (it seems Chrome accepts in-line scripts)
4. Site not loading:
image

### Expected behavior:

Be able to set CSP accordingly with the needs of the spesific installation.

### Actual behavior:

Rocket Chat not working at all.

### Server Setup Information:

- Version of Rocket.Chat Server: The one that are in production at https://open.rocket.chat/

### Client Setup Information

- Desktop App or Browser Version: Firefox 83
- Operating System: Fedora Linux 33

### Additional context - Steps to find the in-line content:

1. Download a Content Security Policy (CSP) Generator for your browser: https://addons.mozilla.org/en-US/firefox/addon/csp-generator/ for Firefox or https://chrome.google.com/webstore/detail/content-security-policy-c/ahlnecfloencbkpfnpljbojmjkfgnmdc for Chrome.
2. Got to https://open.rocket.chat/home and select Start Building Policy in the addon
3. Refresh the site
4. Click all the way through "3. Inline" to find all in-line content. Right now this is how the list looks like:

script-src
Location | Line | Column | Script Sample
-- | -- | -- | --
https://open.rocket.chat/homehttps://open.rocket.chat/f2bd94c70859aaeb5491f476536c4eb930b63cd4.js?meteor_js_resource=true | 44 | 828 | /**  * Implements API class that embeds …
https://open.rocket.chat/homehttps://open.rocket.chat/home | 480 | 1 | __meteor_runtime_config__ = JSON.parse(d…
https://open.rocket.chat/homemoz-extension://abb601c1-84c7-4ace-a913-a75bad9df8e2/data/fingerprint-protection.js | 343 | 25 | try {                         Object.def…
https://open.rocket.chat/homehttps://open.rocket.chat/home | 26 | 1 | window.USE_REST_FOR_DDP_CALLS = true;
https://open.rocket.chat/homemoz-extension://ceaf0c05-e3ac-421e-8862-f992ed239038/onloadwff.js | 71 | 799505 | ;!function(){function e(t,e){return"func…

style-src
Location | Line | Column | Script Sample
-- | -- | -- | --
https://open.rocket.chat/homehttps://open.rocket.chat/f2bd94c70859aaeb5491f476536c4eb930b63cd4.js?meteor_js_resource=true | 219 | 12234 | display:none
https://open.rocket.chat/homehttps://open.rocket.chat/home | 436 | 1 | :root {--rc-color-alert: var(--color-ye…
https://open.rocket.chat/homehttps://open.rocket.chat/f2bd94c70859aaeb5491f476536c4eb930b63cd4.js?meteor_js_resource=true | 5 | 613 | .read-receipt{position:absolute;top:2px;…
https://open.rocket.chat/homehttps://open.rocket.chat/f2bd94c70859aaeb5491f476536c4eb930b63cd4.js?meteor_js_resource=true | 5 | 613 | .toast-title{font-weight:700}.toast-mess…
https://open.rocket.chat/homehttps://open.rocket.chat/home | 0 | 0 | display: none
https://open.rocket.chat/homehttps://open.rocket.chat/f2bd94c70859aaeb5491f476536c4eb930b63cd4.js?meteor_js_resource=true | 5 | 613 | .hljs{display:block;overflow-x:auto;padd…
https://open.rocket.chat/homehttps://open.rocket.chat/f2bd94c70859aaeb5491f476536c4eb930b63cd4.js?meteor_js_resource=true | 426 | 231913 | :root { --content-background-color: #fff…
https://open.rocket.chat/homehttps://open.rocket.chat/home | 0 | 0 | displ

### PR DESCRIPTION
**Pull request by ggazzo**
Closes #19815

## Proposed changes (including videos or screenshots)

Security policies were applied for inline scripts cases. Due to the libraries and components we use it is not possible to disable inline styles and images as they would break Oembeds and other libraries.

basically the inline scripts were moved to a js file

and besides that some suggars syntax like `addScript` and `addStyle` were added, this way the application already takes care of inserting the elements and providing the content automatically.

### CODE REVIEW

File: app/lib/server/startup/settings.js [hunk 1]
```diff
@@ -847,6 +847,12 @@ settings.addGroup('General', function() {
 		type: 'boolean',
 		secret: true,
 	});
+
+	this.add('Enable_CSP', true, {
+		type: 'boolean',
+		secret: true,
```
Comment (sampaiodiego) [inline app/lib/server/startup/settings.js:856]:
```suggestion
```

File: app/ui-master/server/inject.ts [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+	if (!content.trim) {
+		injectIntoHead(key, '');
+		return;
+	}
+	const currentHash = crypto.createHash('sha1').update(content).digest('hex');
+	injectIntoHead(`${ key }.js`, { type: 'JS', tag: `<script id="${ key }" type="text/javascript" src="${ `${ getURL(key) }.js?${ currentHash }` }"></script>`, content });
+};
+
+export const addStyle = (key: string, content: string): void => {
+	if (!content.trim) {
```
Comment (sampaiodiego) [inline app/ui-master/server/inject.ts:84]:
```suggestion
	if (!content.trim()) {
```

File: app/ui-master/server/inject.ts [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+};
+
+WebApp.connectHandlers.use(callback);
+
+export const injectIntoHead = (key: string, value: Injection): void => {
+	headInjections.set(key, value);
+};
+
+export const addScript = (key: string, content: string): void => {
+	if (!content.trim) {
```
Comment (sampaiodiego) [inline app/ui-master/server/inject.ts:75]:
```suggestion
	if (!content.trim()) {
```

File: app/ui-master/server/inject.ts [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+		injectIntoHead(key, '');
+		return;
+	}
+	const currentHash = crypto.createHash('sha1').update(content).digest('hex');
+	injectIntoHead(`${ key }.js`, { type: 'JS', tag: `<script id="${ key }" type="text/javascript" src="${ `${ getURL(key) }.js?${ currentHash }` }"></script>`, content });
+};
+
+export const addStyle = (key: string, content: string): void => {
+	if (!content.trim) {
+		injectIntoHead(key, '');
```
Comment (sampaiodiego) [inline app/ui-master/server/inject.ts:85]:
```suggestion
		injectIntoHead(`${ key }.css`, '');
```

File: app/ui-master/server/inject.ts [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+
+WebApp.connectHandlers.use(callback);
+
+export const injectIntoHead = (key: string, value: Injection): void => {
+	headInjections.set(key, value);
+};
+
+export const addScript = (key: string, content: string): void => {
+	if (!content.trim) {
+		injectIntoHead(key, '');
```
Comment (sampaiodiego) [inline app/ui-master/server/inject.ts:76]:
```suggestion
		injectIntoHead(`${ key }.js`, '');
```

**Review by sampaiodiego — CHANGES_REQUESTED** [review]

**Review by sampaiodiego — APPROVED** [review]

**Comment by DocDocTeam** [comment]
Updated to 3.16. There was an error Content-Security-Policy 
"Refused to load the script 'https://lib.usedesk.ru/secure.usedesk.ru/widget_111111.js' because it violates the following Content Security Policy directive: "script-src 'self' 'unsafe-eval'". Note that 'script-src-elem' was not explicitly set, so 'script-src' is used as a fallback."

While disabled in the general settings Content-Security-Policy.
The documentation is empty https://github.com/RocketChat/docs/blob/master/guides/administration/administration/settings/setup-wizard.md

1. You write "basically the inline scripts were moved to a js file" - in which one?
2. How do I whitelist my script?

**Pull request merged**